hypothesis-python/src/hypothesis/extra/_array_helpers.py

```diff
@@ -596,8 +595,8 @@ def _draw_loop_dimensions(self, data, use=None):
                 # shape-tuple even if it is no longer being added to.
                 # This helps to ensure more stable shrinking behavior.
                 if self.min_dims < dim_count:
-                    use[shape_id] &= cu.biased_coin(
-                        data, 1 - 1 / (1 + self.max_dims - dim)
+                    use[shape_id] &= data.draw_boolean(
+                        p=1 - 1 / (1 + self.max_dims - dim)
```

Comment:
I've applied this, though I won't pretend to fully understand the reasoning (and I am definitely not going to dive into the array extra right now!). 

As I understand it, this is "just" changing the calculation of `p` from `1 - 1 / (1 + self.max_dims - dim)` to `_calc_p_continue(desired_avg=self.max_dims - self.min_dims, max_size=self.max_dims - self.min_dims)`. I can see why `_calc_p_continue` is a very nice utility, but I don't see why changing the particular probability from some `p1` here to some `p2` has a meaningful impact on shrinking. 

I can understand why introducing a draw dependence on a var `x` that can change while shrinking would cause bad shrinking behavior. But `dim` doesn't change while shrinking here; it is fixed, given fixed `self.max_dims`:

```python
        for dim_count in range(1, self.max_dims + 1):
            dim = dim_count - 1
```

I'm probably missing something!

Reply:
To be clear, this change is conceptually _totally unrelated_ to the IR changes - and in fact will be obsoleted by the final stage.  The idea is that we want to decide whether to add another dimension _with constant probability_, so that the choice of whether or not to do so draws the same number of bits under the hood.  That in turn is a principle that makes it easier for shrink passes to recognise that there's common structure and mess around with it in useful ways.

/ok, back out of the rabbithole now 😅

Reply:
aha, I was missing that a dynamic-sized series of bits being drawn is distinctly different from a constant-sized series of bits, even within a single test case. This makes sense - thanks!

Yup, I knew this was an unrelated sidebar - but I was curious and wanted to clear up my understanding 🙂